src/main/kotlin/com/exactpro/th2/codec/xml/XmlPipelineCodec.kt

```diff
@@ -0,0 +1,570 @@
+                            msgStructure!!
+                        ).build()
+                    }
+                }
+            }
+        } catch (e: Exception) {
+            when (e) {
+                is IOException,
+                is SAXException -> {
+                    throw DecodeException("Can not decode message. Can not parse XML. ${rawMessage.toJson()}", e)
+                }
+                else -> throw e
+            }
+        }
+    }
+
+    protected fun findMessageTypes(node: Node): Map<String, List<Node>> {
+        if (!node.hasChildNodes()) {
+            return emptyMap()
+        }
+
+        val rootNode =
+            if (xmlRootTagName != null && node.firstChild.nodeName == xmlRootTagName) node.firstChild else node
+
+        val result = HashMap<String, List<Node>>()
+        messagesTypes.forEach { (type, msgStructure) ->
+            (msgStructure.getXPathExpression()?.let {
+                X_PATH.find(it, rootNode) {
+                    DecodeException(
+                        "Can not execute XPath exception for message type '$type'",
+                        it
+                    )
+                }.toList()
+            } ?: rootNode.childNodes.toList().filter { msgStructure.isValidNode(it) }).also {
+                if (it.isNotEmpty()) {
+                    result[type] = it
+                }
+            }
+        }
+
+        return result
+    }
+
+    protected fun decodeMessageNode(
+        node: Node,
+        direction: Direction,
+        sessionAlias: String,
+        messageStructure: IMessageStructure
+    ): Message.Builder {
+        val msgType = messageStructure.name
+        val messageBuilder = message(msgType, direction, sessionAlias)
+        messageBuilder.metadataBuilder.protocol = this.protocol
+
+        decodeMessageFields(messageBuilder, node, direction, sessionAlias, messageStructure)
+
+        return messageBuilder
+    }
+
+    protected fun decodeMessageFields(
+        messageBuilder: Message.Builder,
+        node: Node,
+        direction: Direction,
+        sessionAlias: String,
+        messageStructure: IMessageStructure
+    ) {
+        val notFoundFields = HashSet<String>()
+
+        messageStructure.fields.forEach { (fieldName, fieldStructure) ->
+            val expression = fieldStructure.getXPathExpression()
+            if (expression != null) {
+                val nodes = X_PATH.find(
+                    expression,
+                    node
+                ) { DecodeException("Can not execute XPath expression for field '$fieldName' in message with message type '${messageStructure.name}'") }
+                    .toList()
+                if (nodes.isNotEmpty()) {
+                    decodeFieldNode(messageBuilder, fieldName, nodes, direction, sessionAlias, fieldStructure)
+                    return@forEach
+                }
+            }
+
+            notFoundFields.add(fieldName)
```

Comment:
maybe rename to fields without XPath attribute and add only fields without the XPath Expression